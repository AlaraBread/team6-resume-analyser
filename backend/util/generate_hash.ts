import { encodeHex } from "@std/encoding";
import { crypto } from "./deps.ts";

// this will need to get saved to a file at some point
const salts: Record<string, string> = {};

export async function generateHash(
	username: string,
	password: string,
): Promise<string> {
	const salt = username in salts
		? salts[username]
		: salts[username] = crypto.randomUUID();
	// Hash the password
	const messageBuffer = new TextEncoder().encode(salt + password);
	const hashBuffer = await crypto.subtle.digest(
		"SHA-256",
		messageBuffer,
	);
	const hashedPassword = encodeHex(new Uint8Array(hashBuffer));

	return hashedPassword;
}
