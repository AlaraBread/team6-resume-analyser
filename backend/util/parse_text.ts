import { officeparser } from "./deps.ts";
import { Buffer } from "node:buffer";

// Extract text from a file
export async function processAndCleanFile(file: File): Promise<string> {
	const data = await officeparser.parseOfficeAsync(
		Buffer.from(await file.arrayBuffer()),
	);

	// Ensure data is a string or convert it to a string
	let extractedText: string;
	if (typeof data === "string") {
		extractedText = data;
	} else if (data && typeof data === "object" && "toString" in data) {
		extractedText = (data as { toString: () => string }).toString();
	} else {
		return Promise.reject(
			"Unexpected data type received from parseOfficeAsync",
		);
	}

	// Normalize whitespace: replace multiple spaces and line breaks with a single space
	const cleanedText = extractedText.replace(/\s+/g, " ").trim();

	return cleanedText;
}
