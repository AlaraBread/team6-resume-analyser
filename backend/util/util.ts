export function dedup<T>(arr: T[]) {
	return [...new Set(arr)];
}
