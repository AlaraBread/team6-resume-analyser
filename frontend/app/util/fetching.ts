import useSWR, { SWRResponse } from "swr";

export const jsonFetcher: (
	...args: Parameters<typeof fetch>
) => Promise<object> = (...args) => fetch(...args).then((res) => res.json());

function getEndpointPath(endpoint: string) {
	// ensure trailing slash in server
	const server = process.env.NEXT_PUBLIC_BACKEND?.endsWith("/")
		? process.env.NEXT_PUBLIC_BACKEND
		: process.env.NEXT_PUBLIC_BACKEND + "/";
	// remove trailing slash from endpoint
	if (endpoint.endsWith("/")) {
		endpoint = endpoint.slice(0, endpoint.length - 1);
	}
	// remove leading slash from endpoint
	if (endpoint.startsWith("/")) {
		endpoint = endpoint.slice(1, endpoint.length);
	}
	return server + endpoint;
}

export type getRequests = {
	"api/hello": {
		response: {
			message: string;
		};
	};
};

export function useBackendGet<T extends keyof getRequests>(
	endpoint: T,
): SWRResponse<getRequests[T]["response"], Error> {
	return useSWR(getEndpointPath(endpoint), jsonFetcher) as SWRResponse<
		getRequests[T]["response"],
		Error
	>;
}

export type postRequests = {
	"api/greeting": {
		request: {
			name: string;
		};
		response: {
			message: string;
		};
	};
};

export async function backendPost<T extends keyof postRequests>(
	endpoint: T,
	data: postRequests[T]["request"],
): Promise<postRequests[T]["response"]> {
	return fetch(getEndpointPath(endpoint), {
		method: "POST",
		body: JSON.stringify(data),
	}).then((response) => response.json()) as Promise<
		postRequests[T]["response"]
	>;
}
