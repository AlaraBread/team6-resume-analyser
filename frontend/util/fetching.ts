import useSWR, { SWRResponse } from "swr";

export const jsonFetcher: (
	...args: Parameters<typeof fetch>
) => Promise<object> = (...args) => fetch(...args).then((res) => res.json());

function getEndpointPath(endpoint: string) {
	let server = process.env.NEXT_PUBLIC_BACKEND ?? "";
	// ensure trailing slash in server
	if (!server.endsWith("/")) {
		server = server + "/";
	}
	// remove trailing slash from endpoint
	if (endpoint.endsWith("/")) {
		endpoint = endpoint.slice(0, endpoint.length - 1);
	}
	// remove leading slash from endpoint
	if (endpoint.startsWith("/")) {
		endpoint = endpoint.slice(1, endpoint.length);
	}
	return server + endpoint;
}

export type getRequests = {
	"api/hello": {
		response: {
			message: string;
		};
	};
};

export function useBackendGet<T extends keyof getRequests>(
	endpoint: T,
): SWRResponse<getRequests[T]["response"], Error> {
	return useSWR(getEndpointPath(endpoint), jsonFetcher) as SWRResponse<
		getRequests[T]["response"],
		Error
	>;
}

export type postRequests = {
	"api/greeting": {
		request: {
			name: string;
		};
		response: {
			message: string;
		};
	};
	"api/job-description": {
		request: {
			description: string;
		};
		response: {
			message: string;
		};
	};
	"api/resume-upload": {
		request: FormData;
		response: {
			message: string;
		};
	};
};

type formPostRequests = {
	[K in keyof postRequests]-?: postRequests[K]["request"] extends FormData
		? K
		: never;
}[keyof postRequests];

type jsonPostRequests = {
	[K in keyof postRequests]-?: postRequests[K]["request"] extends FormData
		? never
		: K;
}[keyof postRequests];

export async function backendPost<T extends jsonPostRequests>(
	endpoint: T,
	data: postRequests[T]["request"],
): Promise<postRequests[T]["response"]> {
	return fetch(getEndpointPath(endpoint), {
		method: "POST",
		body: JSON.stringify(data),
	}).then((response) => response.json()) as Promise<
		postRequests[T]["response"]
	>;
}

export async function backendFormPost<T extends formPostRequests>(
	endpoint: T,
	data: postRequests[T]["request"],
): Promise<postRequests[T]["response"]> {
	return fetch(getEndpointPath(endpoint), {
		method: "POST",
		body: data,
	}).then((response) => response.json()) as Promise<
		postRequests[T]["response"]
	>;
}
