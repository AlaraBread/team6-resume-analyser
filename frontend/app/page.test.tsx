import { act, fireEvent, render, screen } from "@testing-library/react";
import Page from "./page";
import { processEvents } from "util/testing";

// https://jestjs.io/docs/mock-functions
jest.mock("../util/fetching", () => {
	const originalModule = jest.requireActual("../util/fetching");
	return {
		__esModule: true,
		...originalModule,
		useBackendGet: (endpoint: string) => {
			switch (endpoint) {
				case "api/hello":
					return {
						data: {
							text: "Hello from the backend!",
							otherText: "Hello again!",
						},
						error: undefined,
					};
			}
		},
		backendPost: (endpoint: string, data: any) => {
			switch (endpoint) {
				case "api/greeting":
					return Promise.resolve({
						message: "this is a message",
					});
			}
		},
	};
});

it("has an h1 that says resume analyser", () => {
	render(<Page />);
	expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent(
		"resume analyser!",
	);
});

it("should match snapshot", () => {
	const { container } = render(<Page />);
	expect(container).toMatchSnapshot();
});

it("displays the correct information from the backend get", () => {
	render(<Page />);
	expect(screen.getByTestId("backend-example-get").textContent).toEqual(
		'{"text":"Hello from the backend!","otherText":"Hello again!"}',
	);
});

it("displays the correct information from the backend post", async () => {
	render(<Page />);
	await act(async () => {
		fireEvent.click(
			screen.getByRole("button", { name: "send post request" }),
		);
	});
	expect(screen.getByTestId("backend-example-post").textContent).toEqual(
		"this is a message",
	);
});
