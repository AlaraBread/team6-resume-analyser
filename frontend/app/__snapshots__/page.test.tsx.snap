// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div>
  <div
    class="center"
  >
    <h1>
      resume analyser!
    </h1>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <h2
        class="MuiCardHeader-root css-149tc0y-MuiCardHeader-root"
        data-testid="counter-header"
      >
        <div
          class="MuiCardHeader-content css-1jkp9bu-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-16xl4zq-MuiTypography-root"
          >
            count: 0
          </span>
        </div>
      </h2>
      <div
        class="MuiCardContent-root buttons css-1lt5qva-MuiCardContent-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1abvrni-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          more
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1btmbba-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          less
        </button>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <h2
        class="MuiCardHeader-root css-149tc0y-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-content css-1jkp9bu-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-16xl4zq-MuiTypography-root"
          >
            GET
          </span>
        </div>
      </h2>
      <div
        class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
      >
        <p
          data-testid="backend-example-get"
        >
          {"text":"Hello from the backend!","otherText":"Hello again!"}
        </p>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <h2
        class="MuiCardHeader-root css-149tc0y-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-content css-1jkp9bu-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-16xl4zq-MuiTypography-root"
          >
            POST
          </span>
        </div>
      </h2>
      <div
        class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
      >
        <p
          data-testid="backend-example-post"
        />
        <form>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-1cmet5c-MuiInputBase-root-MuiInput-root"
          >
            <input
              class="MuiInputBase-input MuiInput-input css-1yrc8ca-MuiInputBase-input-MuiInput-input"
              name="name"
              placeholder="name"
              type="text"
              value=""
            />
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1abvrni-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            send post request
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
